pingcap/tidb-operator
# wait leaders transfer back after a tikv upgrade

### What problem does this PR solve?


* record leader count in the `status.tikv.store[].leaderCountBeforeUpgrade` before a tikv eviction begin.
* record the time when a tikv leader eviction end in the annotation `tidb.pingcap.com/tikv-evict-leader-end-at` of the pod.
* Before a tikv leader eviction begins, wait for leaders of the previous tikv to transfer back.
  The following cases indicate that the transfer is done:
    1. miss information to decide the result, like: `leaderCountBeforeUpgrade` or `tidb.pingcap.com/tikv-evict-leader-end-at`.
    2. `leaderCountBeforeUpgrade` is less than 200.
    3. timeout, default to 400s.
    4. current leader count  is greater than `leaderCountBeforeUpgrade` * 2/3

### What is changed and how does it work?


### Code changes


### Tests



1. Deploy a TiDB cluster.
2. Upgrade the TiKV.
3. Check that wait for leaders to transfer back before upgrading the next tikv.


### Side effects


### Related changes


### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note

```
